### PR TITLE
Python 3 support for dockerpty

### PR DESCRIPTION
**Pull request by rmohr**
Hi,

the code changes below add full Python 3 support. Sadly I found an issue in Client.attach_socket of the docker-py library.
I created the pull request docker/docker-py#413 to fix this.

All unit tests and integration tests run fine on Python 3.3.2 and Python 2.7.5 for me when patching docker-py with request docker/docker-py#413.

When applying #27, the travis builds for Python 2.7 also succeed.

Best Regards,
Roman

### CODE REVIEW

**Comment by rmohr** [comment]
I have rebased the pull request and issue docker/docker-py#413 is now allso fixed in upstream docker-py.

File: tests/util.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# dockerpty: util.py
+# iockerpty: util.py
```
Comment (dnephin) [inline tests/util.py]: was this edit an accident?

**Comment by dnephin** [comment]
LGTM

File: tests/util.py (hunk 1 above)
Comment (rmohr) [inline tests/util.py]: Yep. Fixed. Thx.

**Comment by rmohr** [comment]
docker-py 0.7.1 is out. So I have updated tox and travis to run the tests for Python3 too.

The travis builds can not run in parallel, that is why I had to set the parallel builds in the Travis repository settings to 1.
Any ideas how to avoid that? Maybe we could run tox in the uml. Tox runs all builds in a sequence. What do you think?

**Comment by rmohr** [comment]
docker-py 0.7.1 is out. So I have updated tox and travis to run the tests for Python3 too.

The travis builds can not run in parallel, that is why I had to set the parallel builds in the Travis repository settings to 1. Maybe this is not an ideal solution. Any suggestions how to solve that

**Comment by rmohr** [comment]
@dnephin Thank you for reviewing the changes!
@aanand What do you say? Is the pull request ready to be merged?

This would really help me to fix this [issue](https://bugzilla.redhat.com/show_bug.cgi?id=1172357) in fedora.

File: tests/unit/test_io.py [hunk 2]
```diff
@@ -137,64 +131,59 @@ class TestDemuxer(object):
...
-            "\x01\x00\x00\x00\x00\x00\x00\x01d",
+            b"\x01\x00\x00\x00\x00\x00\x00\x03foo",
+            b"\x01\x00\x00\x00\x00\x00\x00\x01d",
         ]
-        return StringIO(u''.join(chunks))
-
+        if six.PY3:
+            return BytesIO(bytes().join(chunks))
+        else:
+            return StringIO(u''.join(chunks))
```
Comment (aanand) [inline tests/unit/test_io.py]: This should work: `six.BytesIO(six.binary_type().join(chunks))`

File: features/steps/step_definitions.py [hunk 3]
```diff
@@ -113,22 +114,22 @@ def step_impl(ctx, rows, cols):
 
 @when('I type "{text}"')
 def step_impl(ctx, text):
-    util.write(ctx.pty, text)
+    util.write(ctx.pty, text.encode() if six.PY3 else text)
```
Comment (aanand) [inline features/steps/step_definitions.py]: This should work: `six.binary_type(text)`

File: dockerpty/io.py [hunk 4]
```diff
@@ -201,16 +194,15 @@ def read(self, n=4096):
         if size <= 0:
             return
         else:
-            data = ''
+            data = bytes() if six.PY3 else str()
```
Comment (aanand) [inline dockerpty/io.py]: This should work: `six.binary_type()`

**Comment by aanand** [comment]
Few places where version sniffing can be replaced with `six` utilities. Otherwise, nice. Could you squash it?

File: tests/unit/test_io.py (hunk 2 above)
Comment (rmohr) [inline tests/unit/test_io.py]: done

File: features/steps/step_definitions.py (hunk 3 above)
Comment (rmohr) [inline features/steps/step_definitions.py]: `text.encode()` turned out to be enough

File: dockerpty/io.py (hunk 4 above)
Comment (rmohr) [inline dockerpty/io.py]: done

**Comment by rmohr** [comment]
@aanand Suggestions included and squashed. Thanks for reviewing!

**Comment by aanand** [comment]
Sorry I missed this. There are still a bunch more `bytes() if six.PY3 else str()` - could you replace those all with `six.binary_type()`? Thanks.

**Comment by rmohr** [comment]
Replaced all occurrences.

The commit 04986b5 succeeds [here](https://travis-ci.org/rmohr/dockerpty/builds/50756399) and fails in the [pull request](https://travis-ci.org/d11wtq/dockerpty/builds/50756405) because of the random IO error I mentioned above. It should be enough to rerun the build in Travis.

**Comment by rmohr** [comment]
@aanand  could you take another look? Thx :)

**Comment by aanand** [comment]
LGTM

**Pull request merged**